pkg/util/netlink.go

```diff
@@ -139,3 +152,102 @@ func OnLinkEvent(name string, do func(), stop <-chan struct{}, errcb func(error)
 		}
 	}
 }
+
+// Move an existing macvtap interface from the current netns to the target netns, and rename it..
+// Optionally configure the MAC address of the interface and the link's MTU.
+func ConfigureInterface(currentIfaceName string, newIfaceName string, macAddr *net.HardwareAddr, mtu int, netns ns.NetNS) (*current.Interface, error) {
+	var err error
+
+	macvtapIface, err := netlink.LinkByName(currentIfaceName)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup device %q: %v", currentIfaceName, err)
+	}
+
+	// move the macvtap interface to the pod's netns
+	if err = netlink.LinkSetNsFd(macvtapIface, int(netns.Fd())); err != nil {
+		return nil, fmt.Errorf("failed to move iface %s to the netns %d because: %v", macvtapIface, netns.Fd(), err)
+	}
+
+	var macvtap *current.Interface = nil
+
+	// configure the macvtap iface
+	err = netns.Do(func(_ ns.NetNS) error {
+		defer func() {
+			if err != nil {
+				LinkDelete(currentIfaceName)
+				LinkDelete(newIfaceName)
+			}
+		}()
+
+		if mtu != 0 {
+			if err := netlink.LinkSetMTU(macvtapIface, mtu); err != nil {
+				return fmt.Errorf("failed to set the macvtap MTU for %s: %v", currentIfaceName, err)
+			}
+		}
+
+		if macAddr != nil {
+			if err := netlink.LinkSetHardwareAddr(macvtapIface, *macAddr); err != nil {
+				return fmt.Errorf("failed to add hardware addr to %q: %v", currentIfaceName, err)
+			}
+		}
+
+		renamedMacvtapIface, err := renameInterface(macvtapIface, newIfaceName)
+		if err != nil {
+			return err
+		}
+
+		// set proxy_arp on the interface
+		if err := configureArp(newIfaceName); err != nil {
+			return err
+		}
+
+		if err := netlink.LinkSetUp(renamedMacvtapIface); err != nil {
+			return fmt.Errorf("failed to set macvtap iface up: %v", err)
+		}
+
+		// Re-fetch macvtap to get all properties/attributes
+		// This enables us to report back the MAC address assigned to the macvtap iface
+		// and now that we've handed the macvtap over, update the netns where it runs
+		macvtapIface, err = netlink.LinkByName(newIfaceName)
+		if err != nil {
+			return err
+		}
+
+		macvtap = &current.Interface{
+			Name:    newIfaceName,
+			Mac:     macvtapIface.Attrs().HardwareAddr.String(),
+			Sandbox: netns.Path(),
+		}
```

Comment:
Ditto, as my previous cosmetic comment, with this section. The `current.Interface` is directly related to the cni api and I would prefer it to be built in the main CmdAdd method than buried in an utility method. `ConfigureInterface` could return nothing, and we could have a specific utility method to re-fetch the mac address (only if none was specified).